fix the datetime set in backtracking correction

pbench opened the pull request:
There was a confusion between the time-in-day (`st.arrival_time` and `st.departure_time`) and time-since-beginning-of-dataset (`section.get_out_dt` and `section.get_in_dt`).

So when a section was shorten, its modified endpoint was always set in the first day of the calendar.
We now compute the `base_dt` (that is midnight for the day on which the vj is used) from the section data, and then use this to set the correct datetime in the modified section.

Fix for https://navitia.atlassian.net/browse/NAV-1473

pbench committed:
fix the datetime set in backtracking correction

pbench commented:
I'll try to add some tests

pbench committed:
some tests

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -1408,8 +1408,19 @@ bool shorten_section(navitia::routing::Journey::Section& section,
+                    section.get_out_dt = base_dt + st.arrival_time;
+                } else {
+                    section.get_in_st = &st;
+                    section.get_in_dt = base_dt + st.departure_time;
```
Aren't you supposed to use `st.boarding_time` here instead? (alighting above)

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -1408,8 +1408,19 @@ bool shorten_section(navitia::routing::Journey::Section& section,
              boost::make_iterator_range(vj->stop_time_list.begin() + order.val, vj->stop_time_list.end())) {
             if (can_shorten_at(departures, destinations, st, clockwise)
                 && st.stop_point->stop_area->uri == stop_area_uri && vj != vj_to_skip) {
```
For clarity and perf, It looks like:
1. `vj != vj_to_skip` can be moved in the above loop (the one on vj), with a `continue`
2. the `stop_area` check could be done before, as it is gonna be false more often than `can_shorten_at()`

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -1408,8 +1408,19 @@ bool shorten_section(navitia::routing::Journey::Section& section,
+                // with protection from underflow
+                navitia::DateTime base_dt = 0;
+                if (section.get_in_dt > section.get_in_st->departure_time) {
+                    base_dt = section.get_in_dt - section.get_in_st->departure_time;
```
As discussed:
That `base_dt` processing should always be OK (except maybe for pass-midnight stay-in).
So underflow case is the sign of a bug somewhere else, and we should probably log+default to no-shortening (instead of defaulting to shortening on first day of dataset).

To handle pass-midnight stay-in (not sure it's possible from the rest of the code, but better to be safe) : this `base_dt` would probably be better computed outside of vj's loop.

Test suggestion (clockwise and anti-clockwise): past-midnight during first VJ, past-midnight during second VJ, past-midnight during stay-in. Maybe also a combination of them?
Those tests may be heavy, so feel free to ignore (part of) this suggestion.

stifoon replied:
Good idea, but as we prioritize the migration from Kraken to Loki, let's keep it simple, even if the solution is limited. And let's assume that some cases won't be handled until we switch to Loki !

pbench committed:
smarter conditions

pbench committed:
alighting/boarding time + log error and return on incoherent base_dt

pbench committed:
be less smart

pbench committed:
some warning comments

Commit:
update tyr dependencies to have tyr-tests running

Commit:
Fix handling of anticlockwise backtracking correction inside
section (browse VJs and stop_times reversely)

* improve corner cases functionnaly
* improve code and maintenance
* improve perfs a bit

/!\ anticlockwise on journey (browse sections) is not correctly handled

Commit:
Fix handling of anticlockwise backtracking correction inside
journey (browse sections reversely)

As mutualizing is not possible easily, regroup clockwise handling
on one side and anticlockwise on the other (inline functions)

pbougue reviewed (approved):
Mainly self-review on that one :angel:

pbench commented on source/routing/raptor_api.cpp:
```diff
@@ -1369,97 +1370,87 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
         }
     }
-    LOG4CPLUS_WARN(logger, "Error occurred when modifying backtracking journeys");
+    LOG4CPLUS_ERROR(logger, "Error occurred when modifying backtracking journeys clockwise");
```
This is not an error anymore to arrive here. You will reach the end of the function on every section (apart the last one) that cannot be shortened

pbougue replied:
As discussed: Well-seen, only the last section of the VJ uses last_vj, so this case would happen regularly.
But then on intermediate sections we could search too far and create bugs.

So chose to keep it an error message and instead add a test that we reached section's end stop-time: https://github.com/hove-io/navitia/pull/3817/commits/4c3e911487e0d5193e2421e0c3c4546b6638e01c

pbougue replied:
:sweat_smile: Actually, stop shortening search AFTER checking the end stop-time: https://github.com/hove-io/navitia/pull/3817/commits/3bb9c1374a4309b540560449de066d2b9b7d0a1d

pbench commented on source/routing/raptor_api.cpp:
```diff
@@ -1369,97 +1370,87 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
         }
     }
+
+    LOG4CPLUS_ERROR(logger, "Error occurred when modifying backtracking journeys anticlockwise");
```
Same as above, this is not an error anymore to reach this line.

pbougue replied:
See response above :wink:

pbench commented on source/routing/raptor_api.cpp:
```diff
@@ -1469,15 +1460,39 @@ void modify_backtracking_journeys(RAPTOR::Journeys& journeys,
+
+            size_t section_idx = 0;
+            // browse sections and modify any section that allows shortening
+            for (; section_idx < journey.sections.size(); section_idx++) {
```
For clockwise, you should iterate sections backward and stops at the first section that can be shorten.

Commit:
Stop shorten search once the end of original section is reached

As last_vj is the last one of the journey, not necessary the last one
of the section, early return on vj doesn't protect all section's cases.

Commit:
Interrupt shortening search AFTER testing the last stop of section

And not just before :s

pbench commented:
Ok for me :white_check_mark:
